In [1]:
import numpy as np

In [2]:
np.random.seed(42)
consumption = np.random.randint(200,601, size=120)

In [3]:
consumption
# print(len(consumption))

array([302, 548, 470, 306, 271, 388, 220, 302, 321, 414, 530, 287, 572,
       299, 559, 351, 330, 349, 508, 457, 543, 493, 585, 391, 476, 360,
       513, 221, 452, 435, 544, 248, 258, 369, 387, 470, 389, 374, 250,
       563, 254, 443, 519, 330, 506, 334, 220, 528, 366, 473, 587, 288,
       515, 213, 441, 464, 545, 252, 585, 539, 291, 566, 463, 234, 405,
       280, 249, 559, 587, 201, 589, 253, 305, 459, 509, 390, 417, 243,
       361, 401, 469, 550, 503, 470, 414, 451, 389, 495, 412, 407, 436,
       537, 566, 252, 479, 416, 451, 387, 579, 240, 356, 214, 500, 264,
       544, 526, 208, 543, 328, 335, 262, 338, 280, 591, 362, 488, 578,
       460, 430, 240], dtype=int32)

## Doing the basic statistics

In [4]:
mean = np.mean(consumption)
median = np.median(consumption)
minimum = np.min(consumption)
maximum = np.max(consumption)
std = np.std(consumption)

In [5]:
print('Mean:', mean)
print('Median:', median)
print('Minimum:', minimum)
print('Maximum:', maximum)
print('Standard Deviation:', std)

Mean: 408.4916666666667
Median: 414.0
Minimum: 201
Maximum: 591
Standard Deviation: 115.20235789205397


## Identify The Extremes

In [6]:
highest_day = np.argmax(consumption) + 1
lowest_day = np.argmin(consumption) + 1

print(highest_day)
print(lowest_day)

114
70


In [7]:
print("Highest consumption:", maximum, "kWh")
print("Highest consumption Day is Day", highest_day)
print("Lowest consumption:", minimum, "kWh")
print("Lowest consumption Day is Day", lowest_day)


Highest consumption: 591 kWh
Highest consumption Day is Day 114
Lowest consumption: 201 kWh
Lowest consumption Day is Day 70


## Monthly Splitting

In [8]:
# Calculating the days, 120 days divided by 4 months give us 30 days per month 

month1 = consumption[:30]
month2 = consumption[30:60]
month3 = consumption[60:90]
month4 = consumption[90:120]
# print(month1)
# print(month3)
# print(month4)

In [9]:
print("Month 1", month1)
print("Month 2", month2)
print("Month 3", month3)
print("Month 4", month4)

Month 1 [302 548 470 306 271 388 220 302 321 414 530 287 572 299 559 351 330 349
 508 457 543 493 585 391 476 360 513 221 452 435]
Month 2 [544 248 258 369 387 470 389 374 250 563 254 443 519 330 506 334 220 528
 366 473 587 288 515 213 441 464 545 252 585 539]
Month 3 [291 566 463 234 405 280 249 559 587 201 589 253 305 459 509 390 417 243
 361 401 469 550 503 470 414 451 389 495 412 407]
Month 4 [436 537 566 252 479 416 451 387 579 240 356 214 500 264 544 526 208 543
 328 335 262 338 280 591 362 488 578 460 430 240]


## Monthly Analysis

In [10]:
# For month 1
total_month1 = np.sum(month1)
average_month1 = np.mean(month1)

# month 2
total_month2 = np.sum(month2)
average_month2 = np.mean(month2)

#month 3
total_month3 = np.sum(month3)
average_month3 = np.mean(month3)

# month 4
total_month4 = np.sum(month4)
average_month4 = np.mean(month4)

print(total_month1)

12253


In [11]:
print("Month 1 Total", total_month1)
print("Month 1 Average", average_month1)
print("Month 2 Total", total_month2)
print("Month 2 Average", average_month2)
print("Month 3 Total", total_month3)
print("Month 3 Average", average_month3)
print("Month 4 Total", total_month4)
print("Month 4 Average", average_month4)

Month 1 Total 12253
Month 1 Average 408.43333333333334
Month 2 Total 12254
Month 2 Average 408.46666666666664
Month 3 Total 12322
Month 3 Average 410.73333333333335
Month 4 Total 12190
Month 4 Average 406.3333333333333


## Weekly Analysis

In [12]:
weekly_data = consumption[:119]

weeks = weekly_data.reshape(17,7)
print(weeks)
# print(weekly_data)

[[302 548 470 306 271 388 220]
 [302 321 414 530 287 572 299]
 [559 351 330 349 508 457 543]
 [493 585 391 476 360 513 221]
 [452 435 544 248 258 369 387]
 [470 389 374 250 563 254 443]
 [519 330 506 334 220 528 366]
 [473 587 288 515 213 441 464]
 [545 252 585 539 291 566 463]
 [234 405 280 249 559 587 201]
 [589 253 305 459 509 390 417]
 [243 361 401 469 550 503 470]
 [414 451 389 495 412 407 436]
 [537 566 252 479 416 451 387]
 [579 240 356 214 500 264 544]
 [526 208 543 328 335 262 338]
 [280 591 362 488 578 460 430]]


In [13]:
weekly_total = np.sum(weeks, axis=1)
weekly_average = np.mean(weeks, axis=1)
print("Weekly Total:",weekly_total)
print("Weekly Average:",weekly_average)

Weekly Total: [2505 2725 3097 3039 2693 2743 2803 2981 3241 2515 2922 2997 3004 3088
 2697 2540 3189]
Weekly Average: [357.85714286 389.28571429 442.42857143 434.14285714 384.71428571
 391.85714286 400.42857143 425.85714286 463.         359.28571429
 417.42857143 428.14285714 429.14285714 441.14285714 385.28571429
 362.85714286 455.57142857]


## Overall consumption Trend

In [14]:
cumulative_consumption = np.cumsum(consumption)
# print(cumulative_consumption)

print(cumulative_consumption[:10])


[ 302  850 1320 1626 1897 2285 2505 2807 3128 3542]


## Anomaly Detection

In [15]:
upper_limit = mean + (2 * std)
lower_limit = mean - (2 * std)

print(upper_limit, lower_limit)

638.8963824507746 178.08695088255874


In [16]:
anomalies = (consumption > upper_limit) | (consumption < lower_limit)
anomaly_days = np.where(anomalies)[0] + 1
# print(anomalies)

print(anomaly_days)
print("Anomaly values", consumption[anomalies])

[]
Anomaly values []


## Summary Overview

Looking at the dataset, we'll get to know that the mean of the data is 408.49 and the median is 414 which is still on the same range telling us the data is not having any inconsistency that drag the data up or down. For the total sum of each months the values are kinda close to each other in the first month we get 12253 and month 2 we have 12254 which shows the consistency in the electricty consumption data throughout the months. when dealing with the week we only have 17  weeks out of the 120 days leaving out 1 day out. so the first week has a total sum of consumption of 2505 and the week that has the highest is the last week having 3189. The day that has the highest electricity consumption is day 114 of 591kwh and the lowest day consumption 201kwh of day 70.And for the anomaly there is not consumption that exceed the upper limit and below the lower limit  which tells us no outlier in out dataset 